docker/pandoc-lambda/function/app.py

```diff
@@ -1,14 +1,133 @@
 import boto3
+import io
 import os
 import signal
 import subprocess
 import tempfile
+from docx import Document
+from docx.oxml import OxmlElement, parse_xml
+from lxml import etree
+
+def lift_footnote(doc, footnotes_part, ref, texts, id, author=False):
+    id_att = '{http://schemas.openxmlformats.org/wordprocessingml/2006/main}id'
+    custom_mark_att = '{http://schemas.openxmlformats.org/wordprocessingml/2006/main}customMarkFollows'
+    footnote = OxmlElement("w:footnote")
+    footnote.attrib[id_att] = id
+    doc_insert = OxmlElement("w:footnoteReference")
+    doc_insert.attrib[custom_mark_att] = "1"
+    doc_insert.attrib[id_att] = id
+
+    # Content
+    for t in texts:
+        footnote.append(t)
+
+    # Insert into the footnotes file
+    footnotes_part.element.append(footnote)
+
+    # Insert the reference into the doc
+    ref.insert(1,doc_insert)
+
+def promote_case_footnotes(doc):
+    val_att = '{http://schemas.openxmlformats.org/wordprocessingml/2006/main}val'
+    author_footnotes = {}
+
+    for ref in doc.element.xpath("//*[*/w:rStyle[starts-with(@w:val,'FootnoteReference')]]"):
+        mark_text = ref.text
+        style_node = ref.xpath(".//w:rStyle[starts-with(@w:val,'FootnoteReference')]")[0]
+        node_id = style_node.attrib[val_att][18:]
+        mark_id = f"{node_id}-{mark_text}"
+        if mark_id not in author_footnotes:
+            author_footnotes[mark_id] = {'id':mark_id, 'mark': mark_text, 'refs': [], 'texts': []}
+        style_node.attrib[val_att] = 'FootnoteReference'
+        author_footnotes[mark_id]['refs'].append(ref)
+
+    for text_el in doc.element.xpath("//w:p[w:pPr/w:pStyle[starts-with(@w:val,'FootnoteText')]]"):
+        mark_text = text_el.xpath(".//*[*/w:rStyle]/w:t")[0].text
+        node_id = text_el.xpath(".//w:pStyle[starts-with(@w:val, 'FootnoteText')]/@w:val")[0][13:]
+        mark_id = f'{node_id}-{mark_text}'
+        if mark_id not in author_footnotes:
+            author_footnotes[mark_id] = {'id':mark_id, 'mark': mark_text, 'refs': [], 'texts': []}
+        mark_el = text_el.xpath("./w:pPr/w:pStyle[starts-with(@w:val,'FootnoteText')]")[0]
+        mark_el.attrib[val_att] = 'FootnoteText'
+        for style in text_el.xpath(".//w:pStyle[starts-with(@w:val, 'FootnoteText')]"):
+            style.attrib[val_att] = 'FootnoteText'
+        for style in text_el.xpath(".//w:rStyle[starts-with(@w:val, 'FootnoteRef')]"):
+            style.attrib[val_att] = 'FootnoteReference'
+        author_footnotes[mark_id]['texts'].append([text_el])
+
+
+    case_footnotes = {}
+
+    for ref in doc.element.xpath("//*[*/w:rStyle[starts-with(@w:val,'CaseFootnoteReference')]]"):
+        mark_text = ref.text
+        node = ref.xpath(".//w:rStyle[starts-with(@w:val,'CaseFootnoteReference')]")[0]
+        node_id = node.attrib[val_att][22:]
+        mark_id = f'{node_id}-{mark_text}'
+        if mark_id not in case_footnotes:
```

Comment:
(Same question: if `mark_id in author_footnotes` is pathological, maybe we should log when they go by.)

Reply:
This will happen when someone elides a footnote mark in the text, but not the footnote text (which is probably unintended, but expected. Minimizing these would be one of the advantages of a programmatic footnote Ellison option on the UI.

Reply:
Ah, gotcha!